sklearn/ensemble/_forest.py

```diff
@@ -507,53 +585,63 @@ def __init__(self,
-                                                  check_input=False)
+        Parameters
+        ----------
+        tree : DecisionTreeClassifier object
+            A single decision tree classifier.
+        X : ndarray of shape (n_samples, n_features)
+            The OOB samples.
 
-            if self.n_outputs_ == 1:
-                p_estimator = [p_estimator]
+        Returns
+        -------
+        y_pred : ndarray of shape (n_samples, n_classes, n_outputs)
+            The OOB associated predictions.
+        """
+        y_pred = tree.predict_proba(X, check_input=False)
+        y_pred = np.array(y_pred, copy=False)
+        if y_pred.ndim == 2:
+            # binary and multiclass
+            y_pred = y_pred[..., np.newaxis]
+        else:
+            # Roll the first `n_outputs` axis to the last axis. We will reshape
+            # from a shape of (n_outputs, n_samples, n_classes) to a shape of
+            # (n_samples, n_classes, n_outputs).
+            y_pred = np.rollaxis(y_pred, axis=0, start=3)
+        return y_pred
 
-            for k in range(self.n_outputs_):
-                predictions[k][unsampled_indices, :] += p_estimator[k]
+    @staticmethod
+    def _score_oob_predictions(y_true, oob_pred):
+        """Compute the average accuracy for the OOB samples.
 
-        for k in range(self.n_outputs_):
-            if (predictions[k].sum(axis=1) == 0).any():
-                warn("Some inputs do not have OOB scores. "
-                     "This probably means too few trees were used "
-                     "to compute any reliable oob estimates.")
+        Parameters
+        ----------
+        y_true : ndarray of shape (n_samples,) or (n_samples, n_outputs)
+            The true target.
+        oob_pred : ndarray of shape (n_samples, n_classes, n_outputs)
+            The OOB predictions.
 
-            decision = (predictions[k] /
-                        predictions[k].sum(axis=1)[:, np.newaxis])
-            oob_decision_function.append(decision)
-            oob_score += np.mean(y[:, k] ==
-                                 np.argmax(predictions[k], axis=1), axis=0)
+        Returns
+        -------
+        oob_score : float
+            The average accuracy score.
+        """
+        return accuracy_score(y_true, np.argmax(oob_pred, axis=1))
 
-        if self.n_outputs_ == 1:
-            self.oob_decision_function_ = oob_decision_function[0]
-        else:
-            self.oob_decision_function_ = oob_decision_function
+    def _set_oob_score_and_attributes(self, X, y):
+        """Compute and set the OOB score and attributes.
 
-        self.oob_score_ = oob_score / self.n_outputs_
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            The data matrix.
+        y : ndarray of shape (n_samples, n_outputs)
+            The target matrix.
+        """
+        self.oob_decision_function_ = super()._set_oob_score(X, y).squeeze()
```

Comment:
I think we should pass the `axis=` argument to only squeeze the axis for `n_targets=1` explicitly.